ext/openssl/ossl_ssl.c

```diff
@@ -441,6 +441,60 @@ ossl_sslctx_session_new_cb(SSL *ssl, SSL_SESSION *sess)
     return 0;
 }
 
+#if OPENSSL_VERSION_NUMBER >= 0x10101000 && !defined(LIBRESSL_VERSION_NUMBER)
+/*
+ * It is only compatible with OpenSSL >= 1.1.1. Even if LibreSSL implements
+ * SSL_CTX_set_keylog_callback() from v3.4.2, it does nothing (see
+ * https://github.com/libressl-portable/openbsd/commit/648d39f0f035835d0653342d139883b9661e9cb6).
+ */
+
+static VALUE
+ossl_call_keylog_cb(VALUE ary)
+{
+    VALUE ssl_obj, sslctx_obj, cb;
+
+    Check_Type(ary, T_ARRAY);
+    ssl_obj = rb_ary_entry(ary, 0);
+    sslctx_obj = rb_attr_get(ssl_obj, id_i_context);
+
+    cb = rb_attr_get(sslctx_obj, id_i_keylog_cb);
+    if (NIL_P(cb)) return Qnil;
+
+    return rb_funcallv(cb, id_call, 1, &ary);
+}
+
+static VALUE
+rb_str_new_cstr_stub(VALUE str)
+{
+    return rb_str_new_cstr((const char *)str);
+}
+
+
+static void
+ossl_sslctx_keylog_cb(const SSL *ssl, const char *line)
+{
+    VALUE ary, ssl_obj;
+    int state = 0;
+
+    OSSL_Debug("SSL keylog callback entered");
+
+    ssl_obj = (VALUE)SSL_get_ex_data(ssl, ossl_ssl_ex_ptr_idx);
+
+    ary = rb_ary_new2(2);
+    rb_ary_push(ary, ssl_obj);
+    rb_ary_push(ary, rb_protect(rb_str_new_cstr_stub, (VALUE)line, &state));
+    if (state) {
+        rb_ivar_set(ssl_obj, ID_callback_state, INT2NUM(state));
+        return;
+    }
```

Comment:
rb_ary_new2() is also unsafe to use here for the same reason as rb_str_new{2,_cstr}(), it can potentially raise NoMemoryError which is basically a longjmp(). rb_protect() can be used to intercept it but makes code ugly, as you can see.

Sorry that I wasn't clear. We don't necessarily have to use Ruby Array here, so I think the simplest way is to make a temporary `void *[2]` on the stack to store `ssl_obj` and `line` and give it to `ossl_call_keylog_cb()`.

Reply:
Thanks for your review. I think I will need some guidance here, I'm a bit lost. I tried to follow the same logic than the callbacks follow in `ossl_ssl.c` (`ssl_servername_cb`, `ossl_sslctx_session_get_cb`, `ossl_sslctx_session_remove_cb`, etc). They all use Ruby Array, calling `rb_ary_new2()` without being protected with `rb_protect()`. So, at this point, I'm not sure how to implement the whole callback logic without Ruby Array. Sorry, I'm new to Ruby native development.

So, I see two options:
1. wrapping `rb_ary_new2()` around `rb_protect()` the same way  it is done with `rb_str_new_cstr()`.
2. using a temporary `void *[2]` as you suggested (I tried, but couldn't get rid of a segfault). Maybe you can point me to an example in the code?

Reply:
An example with `VALUE[2]` (it's guaranteed `VALUE` and a pointer type have the same size and are interchangeable): https://github.com/ruby/openssl/blob/f5b82e814ba95f17ba907d107aa9f0bfa93e52c2/ext/openssl/ossl_pkey.c#L238-L257

Another example that uses a temporary struct to pass three values. https://github.com/ruby/openssl/blob/f5b82e814ba95f17ba907d107aa9f0bfa93e52c2/ext/openssl/ossl_x509store.c#L48-L106

Whether to use a `void *[2]`/`VALUE[2]` or a temporary struct is purely a preference.

> I tried to follow the same logic than the callbacks follow in `ossl_ssl.c` (`ssl_servername_cb`, `ossl_sslctx_session_get_cb`, `ossl_sslctx_session_remove_cb`, etc). They all use Ruby Array, calling `rb_ary_new2()` without being protected with `rb_protect()`.

This is actually a bug that needs to be fixed. :disappointed: 

> So, I see two options:

Yes, both can work. However, I think we need less code with the latter.